src/preferArrayLiteralRule.ts

```diff
@@ -47,18 +86,40 @@ export class Rule extends Lint.Rules.AbstractRule {
 
         ruleOptions.forEach((opt: unknown) => {
             if (isObject(opt)) {
-                value = opt['allow-type-parameters'] === true;
+                allowSizeArgument = opt['allow-size-argument'] === true;
+                allowTypeParameters = opt['allow-type-parameters'] === true;
             }
         });
 
         return {
-            allowTypeParameters: value
+            allowSizeArgument,
+            allowTypeParameters
         };
     }
 }
 
-function walk(ctx: Lint.WalkContext<Options>) {
-    const { allowTypeParameters } = ctx.options;
+function walk(ctx: Lint.WalkContext<Options>, checker: ts.TypeChecker | undefined) {
+    const { allowTypeParameters, allowSizeArgument } = ctx.options;
+    function checkExpression(type: InvocationType, node: ts.CallExpression | ts.NewExpression): void {
+        const functionName = AstUtils.getFunctionName(node);
+        if (functionName === 'Array' && inRestrictedNamespace(node)) {
+            const callArguments = node.arguments;
+            if (!allowSizeArgument || !callArguments || callArguments.length !== 1) {
+                const failureString = Rule.getReplaceFailureString(type, node.getText());
+                ctx.addFailureAt(node.getStart(), node.getWidth(), failureString);
+            } else {
+                // When typechecker is not available - allow any call with single expression
+                if (checker) {
+                    const argument = callArguments[0];
+                    const argumentType = checker.getTypeAtLocation(argument);
+                    if (!tsutils.isTypeAssignableToNumber(checker, argumentType) || argument.kind === ts.SyntaxKind.SpreadElement) {
```

Comment:
It was unexpected for me that when I used spread of number array type there was no error, however there was an error if I used array of string elements. Added additional check where type comparison wasn't enough.

@JoshuaKGoldberg do you think that check can be left here or should it be moved to previous condition (`if (!allowSizeArgument || !callArguments...`)? If moved - spread element in argument position will be always restricted which is also seems logical.

Reply:
That seems reasonable. A little unexpected (perhaps a comment is warranted?)

Reply:
I'll add comment.
During offline discussion decided to leave checks in current state.